### DISCUSSION TITLE
best practise: handling paging (with nextlink in the response header)

### DISCUSSION THREAD

**Discussion by michabbb** [saloonphp/saloon#121] [Q&A] [answered]
hi there,
i found this package and it sounds very interesting.
but the first thing I was unable to find is: how would you do handle paging if
the API you are talking with returns the next page in the header and I would
like to get all pages and not only a specific page ?
so in my code I would like to have:

```php
$skd->getAllProducts()
```

and this requests parses the header of the first response and if there is a "nextLink" I have to make while loop
and call all other pages, till there is no longer a "nextLink" in the header.

maybe someone can provide a small snippet or give me an example how this "should be" done with this package.

thanks !!!

**Comment by Sammyjo20** [comment]
Hey there @michabbb!

I recently had this question come up for me also, and during my research, I found this really good article where you can use PHP generators. I've linked the article below for credit, but I've modified the example for your use case, hope It helps. I also have assumed that you don't have an API rate limit to worry about :) 

Within your SDK connector, create a method and use a generator to make a request to the API.

```php
public function all($page = 1): Generator
{
    do {
        $request = new SomeRequest;
        $request->addQuery('page', $page); // Or use limit/offset

        // 'offset' => ($page - 1) * 20,
        // 'limit' => 20,

        $response = $this->send($request);
        $results = $response->json('results');

        foreach ($results as $result) {
            yield $result;
        }

        $page++;
    } while ($response['next'] !== null);
}
```

Then in your user-land code you can call the method in something to iterate over and voila!

```php
foreach($sdk->all() as $result) {
     //
}
```

In this example 

Here is the article, it's really useful!

https://bannister.me/blog/using-generators-for-pagination

**Reply by michabbb** [reply]
cool, thanks a lot, very helpful! 👍🙏

**Comment by bumbummen99** [comment]
I created a PaginationResponse custom response type according to the documentation on creating a response. It checks for next based on current request page parameter and the availablility of ``next`.

**Reply by michabbb** [reply]
@bumbummen99 would you mind sharing your code for this? I am new to this package and trying to find out if this makes life easier or not, right now, I only see complications, that's why I am asking here 😏

**Reply by bumbummen99** [reply]
@michabbb I wrote up a small post on how to do it:
https://skyraptor.eu/blog/creating-an-api-integration-using-the-laravel-saloon-package

@Sammyjo20 Maybe something similar can be added to the package as an abstract class?

**Reply by michabbb** [reply]
@bumbummen99 very nice, thank you for your time 👍

**Comment by michabbb** [comment]
found another example in one of the showcases:

https://github.com/kilianweisl/hellocash-php-sdk/blob/main/src/Client/Resources/ArticleResource.php#L15

**Comment by Sammyjo20** [comment, marked as answer]
Hey @michabbb 

I've created an example SDK using Saloon v2 with this "paginator" method built in, I tested it against the PokeAPI and it works great 🤠 I'm going to use this example in a few of my projects, so thanks for asking the question!

https://github.com/Sammyjo20/pokeapi-sdk

Paginator method: https://github.com/Sammyjo20/pokeapi-sdk/blob/main/src/Pokeapi.php#L54

You can install this SDK into your application or download it locally to play around with the paginator, I also provided a property to yield the responses rather than the results too.

**Reply by michabbb** [reply]
nice, thanks a lot 👍examples are always good to understand things 😏
